runtime/android/core/src/org/xwalk/core/XWalkUpdater.java

```diff
@@ -148,42 +245,50 @@ public void run() {
      *         doesn't need to be updated, or the Crosswalk runtime has not been initialized yet.
      */
     public boolean updateXWalkRuntime() {
-        if (mDialogManager.isShowingDialog()) return false;
+        if (mIsDownloading || mDialogManager.isShowingDialog()) return false;
 
         int status = XWalkLibraryLoader.getLibraryStatus();
         if (status == XWalkLibraryInterface.STATUS_PENDING ||
                 status == XWalkLibraryInterface.STATUS_MATCH) return false;
 
-        Log.d(TAG, "Update the Crosswalk runtime with status " + status);
-        mDialogManager.showInitializationError(status, mCancelCommand, mDownloadCommand);
-        return true;
-    }
-
-    /**
-     * Dismiss the dialog showing and waiting for user's input.
-     *
-     * @return Return false if no dialog is being displayed, true if dismissed the showing dialog.
-     */
-    public boolean dismissDialog() {
-        if (!mDialogManager.isShowingDialog()) return false;
-        mDialogManager.dismissDialog();
+        if (isDownloadMode()) {
+            if (mXWalkApkUrl == null) {
+                mXWalkApkUrl = getAppMetaData(META_XWALK_APK_URL);
+                if (mXWalkApkUrl == null) {
+                    Log.e(TAG, "Please specify xwalk_apk_url in meta-data");
+                    Assert.fail();
+                }
+                Log.d(TAG, "Crosswalk APK download URL: " + mXWalkApkUrl);
+            }
+            XWalkLibraryLoader.startDownload(new BackgroundListener(), mActivity, mXWalkApkUrl);
+        } else {
+            Log.d(TAG, "Update the Crosswalk runtime with status " + status);
+            mDialogManager.showInitializationError(status, mCancelCommand, mDownloadCommand);
+        }
         return true;
     }
 
-    /**
-     * Set the download URL of the Crosswalk runtime. By default, the updater will get the URL from
-     * the Android manifest.
-     *
-     * @param url The download URL.
-     */
-    public void setXWalkApkUrl(String url) {
-        mXWalkApkUrl = url;
+    private boolean isDownloadMode() {
```

Comment:
This is needed to differentiate the silent-download and explicit-download-installation since we need to make updateXWalkRuntime() works as expected in these two cases.